src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/SignatureHelper.cs

```diff
@@ -71,11 +72,76 @@ internal static BlobBuilder MethodSignatureEncoder(ModuleBuilderImpl module, Typ
         }
 
         private static void WriteSignatureForType(SignatureTypeEncoder signature, Type type, ModuleBuilderImpl module)
+        {
+            if (type.IsArray)
+            {
+                Type elementType = type.GetElementType()!;
+                int rank = type.GetArrayRank();
+                if (rank == 1)
+                {
+                    WriteSignatureForType(signature.SZArray(), elementType, module);
+                }
+                else
+                {
+                    signature.Array(out SignatureTypeEncoder elTypeSignature, out ArrayShapeEncoder arrayEncoder);
+                    WriteSimpleSignature(elTypeSignature, elementType, module);
+                    arrayEncoder.Shape(type.GetArrayRank(), ImmutableArray.Create<int>(), ImmutableArray.Create<int>(new int[rank]));
+                }
+            }
+            else if (type.IsPointer)
+            {
+                WriteSignatureForType(signature.Pointer(), type.GetElementType()!, module);
+            }
+            else if (type.IsByRef)
+            {
+                signature.Builder.WriteByte((byte)SignatureTypeCode.ByReference);
+                WriteSimpleSignature(signature, type.GetElementType()!, module);
```

Comment:
```suggestion
                WriteSignatureForType(signature, type.GetElementType()!, module);
```
This is necessary to handle things like `ref int[]`.